ground/src/main/java/com/google/android/ground/ui/datacollection/tasks/AbstractTaskFragment.kt

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.ground.ui.datacollection.tasks
+
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.hilt.navigation.fragment.hiltNavGraphViewModels
+import com.google.android.ground.R
+import com.google.android.ground.model.submission.TaskData
+import com.google.android.ground.ui.common.AbstractFragment
+import com.google.android.ground.ui.datacollection.DataCollectionViewModel
+import com.google.android.ground.ui.datacollection.components.ButtonAction
+import com.google.android.ground.ui.datacollection.components.TaskButton
+import com.google.android.ground.ui.datacollection.components.TaskView
+import java.util.*
+import kotlin.properties.Delegates
+
+abstract class AbstractTaskFragment<T : AbstractTaskViewModel> :
+  AbstractFragment(), TaskFragment<T> {
+
+  protected val dataCollectionViewModel: DataCollectionViewModel by
+    hiltNavGraphViewModels(R.id.data_collection)
+
+  private val buttons: EnumMap<ButtonAction, TaskButton> = EnumMap(ButtonAction::class.java)
+  private lateinit var taskView: TaskView
+  override lateinit var viewModel: T
+
+  /** Position of the task in the Job's sorted task list. Used for instantiating the [viewModel]. */
+  override var position by Delegates.notNull<Int>()
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    if (savedInstanceState != null) {
+      position = savedInstanceState.getInt(TaskFragment.POSITION)
+    }
+  }
+
+  override fun onSaveInstanceState(outState: Bundle) {
+    super.onSaveInstanceState(outState)
+    outState.putInt(TaskFragment.POSITION, position)
+  }
+
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View? {
+    super.onCreateView(inflater, container, savedInstanceState)
+
+    @Suppress("UNCHECKED_CAST")
+    viewModel = dataCollectionViewModel.getTaskViewModel(position) as T
```

Comment:
The viewModel needs to be set up onAttach, not in onViewCreated.

Reply:
Done